### PR TITLE
Feature coppa

### PR DESCRIPTION
**Pull request by RagnarR**
* Add COPPA mode in `PubnativeRequest`

### CODE REVIEW

File: library/src/main/java/net/pubnative/library/request/PubnativeRequest.java [hunk 1]
```diff
@@ -45,17 +45,19 @@
...
+import java.util.TreeMap;
 
 public class PubnativeRequest implements PubnativeHttpRequest.Listener,
                                          AdvertisingIdClient.Listener {
 
     private static         String               TAG                = PubnativeRequest.class.getSimpleName();
     protected static final String               BASE_URL           = "http://api.pubnative.net/api/v3/native";
     protected              Context              mContext           = null;
-    protected              Map<String, String>  mRequestParameters = new HashMap<String, String>();
+    protected              Map<String, String>  mRequestParameters = new TreeMap<String, String>();
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/request/PubnativeRequest.java]: why was this changed to TreeMap?

File: library/src/main/java/net/pubnative/library/request/PubnativeRequest.java [hunk 2]
```diff
@@ -389,7 +398,7 @@ public void onPubnativeHttpRequestFail(PubnativeHttpRequest request, Exception e
     public void onAdvertisingIdClientFinish(AdvertisingIdClient.AdInfo adInfo) {
 
         Log.v(TAG, "onAdvertisingIdClientFinish");
-        if (adInfo != null && !adInfo.isLimitAdTrackingEnabled()) {
+        if (adInfo != null && !adInfo.isLimitAdTrackingEnabled() && !mIsCoppaMode) {
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/request/PubnativeRequest.java]: I would try extracting the coppa flag from the parameters instead of using another variable, you have it already set in the parameters, let's just create a private method `Boolean isCoppaModeEnabled()` that checks for the coppa flag value in the parameters

File: library/src/main/java/net/pubnative/library/request/PubnativeRequest.java [hunk 3]
```diff
@@ -222,6 +225,12 @@ public void setTestMode(boolean enabled) {
         setParameter(Parameters.TEST, enabled ? "1" : "0");
     }
 
+    public void setCoppaMode(boolean enabled) {
+        Log.v(TAG, "setCoppaMode");
+        mIsCoppaMode = enabled;
+        setParameter(Parameters.COPPA, enabled ? "1" : "0" );
+    }
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/request/PubnativeRequest.java]: Since this is so specific, I would rely on setParameter, actually adding coppa to Parameters.Coppa is enough as far as I see this

**Comment by cerberillo** [comment]
Let's please also add some tests to ensure that whenever coppa flag is 1 you don't fill defaults lat long and so on

File: library/src/main/java/net/pubnative/library/request/PubnativeRequest.java (hunk 1 above)
Comment (RagnarR) [inline library/src/main/java/net/pubnative/library/request/PubnativeRequest.java]: Sorry, forgot to remove - try to check performance

File: library/src/main/java/net/pubnative/library/request/PubnativeRequest.java (hunk 2 above)
Comment (RagnarR) [inline library/src/main/java/net/pubnative/library/request/PubnativeRequest.java]: make sense 👍

File: library/src/main/java/net/pubnative/library/request/PubnativeRequest.java (hunk 3 above)
Comment (RagnarR) [inline library/src/main/java/net/pubnative/library/request/PubnativeRequest.java]: We need to add the possibility to enable this mode in Medation for the customer. It easier has a method, then describes which parameter and which value you need to set. And I decide to implement it because, as you said, it very specific.

File: library/src/main/java/net/pubnative/library/request/PubnativeRequest.java [hunk 4]
```diff
@@ -45,6 +45,7 @@
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
+import java.util.TreeMap;
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/request/PubnativeRequest.java]: Can we remove the unused import?

File: library/src/main/java/net/pubnative/library/request/PubnativeRequest.java [hunk 5]
```diff
@@ -222,6 +224,23 @@ public void setTestMode(boolean enabled) {
         setParameter(Parameters.TEST, enabled ? "1" : "0");
     }
 
+    public void setCoppaMode(boolean enabled) {
+        Log.v(TAG, "setCoppaMode");
+        setParameter(Parameters.COPPA, enabled ? "1" : "0" );
+    }
+
+    public boolean isCoppaModeEnabled() {
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/request/PubnativeRequest.java]: Let's make this private or protected, not public

File: library/src/main/java/net/pubnative/library/request/PubnativeRequest.java [hunk 6]
```diff
@@ -222,6 +224,23 @@ public void setTestMode(boolean enabled) {
...
+        boolean result = false;
+
+        if (mRequestParameters.containsKey(Parameters.COPPA)) {
+            if (mRequestParameters.get(Parameters.COPPA).equalsIgnoreCase("1")) {
+                result = true;
+            }
+        }
```
Comment (cerberillo) [inline library/src/main/java/net/pubnative/library/request/PubnativeRequest.java]: you could also do this cleaner:
```java
private boolean isCoppaModeEnabled() {
  boolean result = false;
  String coppa = mRequestParameters.get(Parameters.COPPA);
  if (!TextUtils.isEmpty(coppa)) {
    result = coppa.equalsIgnoreCase("1");
  }
  return result
}
```

Comment (RagnarR) [inline library/src/main/java/net/pubnative/library/request/PubnativeRequest.java]: Agree

File: library/src/test/java/net/pubnative/library/request/PubnativeRequestTest.java [hunk 7]
```diff
@@ -229,4 +229,17 @@ public void testOnErrorResponseFromRequestManager() {
         request.onPubnativeHttpRequestFail(null, error);
         verify(listener, times(1)).onPubnativeRequestFailed(eq(request), eq(error));
     }
+
+    @Test
+    public void enableCoppaMode_disableParams_pass() {
```
Comment (cerberillo) [inline library/src/test/java/net/pubnative/library/request/PubnativeRequestTest.java]: naming should be `methodTested_conditions_results` so this should be `setCoppaMode_enable_shouldAvoidSettingSomeDefaultParameters`

**Comment by cerberillo** [comment]
apart from the test naming all around looks quite nice, so let's merge as soon as you modified it
